src/qutip_qip/circuit/text_renderer.py

```diff
@@ -0,0 +1,497 @@
+"""
+Module for rendering a quantum circuit in text format.
+"""
+
+from math import ceil
+from typing import List
+
+from .base_renderer import BaseRenderer, StyleConfig
+from ..operations import Gate, Measurement
+from . import QubitCircuit
+
+__all__ = [
+    "TextRenderer",
+]
+
+
+class TextRenderer(BaseRenderer):
+    """
+    A class to render a quantum circuit in text format.
+    """
+
+    def __init__(self, qc: QubitCircuit, **style):
+
+        # user defined style
+        style = {} if style is None else style
+        style["gate_margin"] = 0
+        self.style = StyleConfig(**style)
+
+        super().__init__(self.style)
+        self._qc = qc
+        self._qwires = qc.N
+        self._cwires = qc.num_cbits
+        self._layer_list = {i: [] for i in range(self._qwires + self._cwires)}
+
+        self._render_strs = {
+            "top_frame": ["  "] * (self._qwires + self._cwires),
+            "mid_frame": ["──"] * self._qwires + ["══"] * self._cwires,
+            "bot_frame": ["  "] * (self._qwires + self._cwires),
+        }
+
+    def _adjust_layer(self, wire_list: List[int], xskip: int):
+        """
+        Adjust the layers by filling the empty spaces with respective characters
+        """
+
+        for wire in wire_list:
+            self._render_strs["top_frame"][wire] += " " * (
+                xskip - len(self._render_strs["top_frame"][wire])
+            )
+            self._render_strs["bot_frame"][wire] += " " * (
+                xskip - len(self._render_strs["bot_frame"][wire])
+            )
+            if wire < self._qwires:
+                self._render_strs["mid_frame"][wire] += "─" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+            else:
+                self._render_strs["mid_frame"][wire] += "═" * (
+                    xskip - len(self._render_strs["mid_frame"][wire])
+                )
+
+    def _add_wire_labels(self):
+        """
+        Add wire labels to the circuit
+        """
+
+        default_labels = [f"q{i}" for i in range(self._qwires)] + [
+            f"c{i}" for i in range(self._cwires)
+        ]
+        for i, label in enumerate(default_labels):
+            self._render_strs["mid_frame"][i] = (
+                f" {label} :" + self._render_strs["mid_frame"][i]
+            )
+            self._render_strs["top_frame"][i] = " " * (len(f" {label} :") + 2)
+            self._render_strs["bot_frame"][i] = " " * (len(f" {label} :") + 2)
+
+        self._layer_list = {
+            i: [len(self._render_strs["mid_frame"][i])]
+            for i in range(self._qwires + self._cwires)
+        }
```

Comment:
What does `_layer_list` actually save? I assume `_layer_list[i]` is a list of length for each element for a wire? Maybe document it somewhere.i

Reply:
Dict was used to address the classical wire issue with measurement joint, though it was handled differently. So, its a list.

Currently, it shifts the entire wire by a space for `q10` because of extra digit.  Will make change to add padding to make all label len same.